### PR TITLE
add network magic to ContractParametersContext

### LINKED ISSUE #2393: Add network magic to ContractParametersContext
**Summary or problem description**
After #2314 we need network magic to sign things and at the moment we do have full `ProtocolSettings` for that in `Wallet`. There are two problems with this approach:
 * NEP-6 wallets are network-agnostic
   `Wallet` is not a direct NEP-6 implementation of course, but it's strongly related to NEP-6, so network magic is something out of its scope, it's a part of signing context, not wallet. Other implementations may not have network magic in the wallet.
 * multisignatures can use different networks
   Consider classic multisignature with `ContractParametersContext` exchange in JSON format, serialized `ContractParametersContext` doesn't contain network magic and this means that users can easily sign transaction for different networks without noticing it.

Combination of these issues creates an interoperability problem for multisignature signing (operating with serialized `ContractParametersContext`) between various wallet implementations.

**Do you have any solution you want to propose?**
Add network magic to `ContractParametersContext` and serialize it to/from JSON. It can then be used to sign things with `Wallet` directly or by checking that current `Wallet` has the same magic as `ContractParametersContext`.

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- Other: Wallet

### PR DESCRIPTION
**Pull request by Qiao-Jin**
Close #2393

### CODE REVIEW

File: src/neo/SmartContract/ContractParametersContext.cs [hunk 1]
```diff
@@ -325,6 +332,7 @@ public JObject ToJson()
             json["items"] = new JObject();
             foreach (var item in ContextItems)
                 json["items"][item.Key.ToString()] = item.Value.ToJson();
+            json["magic"] = Magic;
```
Comment (roman-khimov) [inline src/neo/SmartContract/ContractParametersContext.cs:335]: Not critical, but it's called `network` in neo-go already.

File: src/neo/Wallets/Wallet.cs [hunk 2]
```diff
@@ -593,6 +593,7 @@ public long CalculateNetworkFee(DataCache snapshot, Transaction tx)
         /// <returns><see langword="true"/> if the signature is successfully added to the context; otherwise, <see langword="false"/>.</returns>
         public bool Sign(ContractParametersContext context)
         {
+            if (context.Magic != ProtocolSettings.Magic) return false;
```
Comment (shargon) [inline src/neo/Wallets/Wallet.cs:596]: The idea it's avoid this, isn't it?

Comment (Qiao-Jin) [reply src/neo/Wallets/Wallet.cs:596]: I think we need to check whether wallet has the same magic as ContractParametersContext, as this is its designed functionality

File: src/neo/SmartContract/ContractParametersContext.cs (hunk 1 above)
Comment (Qiao-Jin) [reply src/neo/SmartContract/ContractParametersContext.cs]: fixed

File: src/neo/Wallets/Wallet.cs (hunk 2 above)
Comment (shargon) [reply src/neo/Wallets/Wallet.cs:596]: I think that we should remove this check, what do you think @neo-project/core ?

File: src/neo/SmartContract/ContractParametersContext.cs [hunk 3]
```diff
@@ -62,6 +62,11 @@ public JObject ToJson()
         /// </summary>
         public readonly DataCache Snapshot;
 
+        /// <summary>
+        /// The magic number of the network.
+        /// </summary>
+        public readonly uint Magic;
```
Comment (erikzhang) [inline src/neo/SmartContract/ContractParametersContext.cs:68]: Network?

**Review by shargon — APPROVED** [review]

**Review by shargon — APPROVED** [review]

File: src/neo/SmartContract/ContractParametersContext.cs (hunk 3 above)
Comment (erikzhang) [reply src/neo/SmartContract/ContractParametersContext.cs]: I mean only modify this field.

Comment (shargon) [reply src/neo/SmartContract/ContractParametersContext.cs]: I know, but you don't think that it's better to call it network in other places?

**Comment by Qiao-Jin** [comment]
ping

**Comment by superboyiii** [comment]
Test **PASS**
I think it could be merged. I'll adjust neo-node and neo-modules.
image
image
@shargon @erikzhang

**Review by superboyiii — APPROVED** [review]

**Pull request merged**